### ISSUE TITLE
TVAE resolves one-hot blocks by their largest entry at sampling time, so minority categories are under-generated and rare ones never appear

### ISSUE THREAD

**Issue by sorayapnblm** [sdv-dev/CTGAN#514]
### Environment Details

Please indicate the following details about the environment in which you found the bug:

* CTGAN version: 0.12.1
* Python version: 3.13.15
* torch: 2.11.0+cpu
* Platform: Google Colab

**Note:** the code quoted below is from `main`. I also checked that `TVAE.sample`,
`_loss_function`, `DataTransformer._inverse_transform_discrete` and
`DataTransformer._inverse_transform_continuous` are character for character the same at the
`v0.12.1` tag I ran.

### Error Description

Hi, I ran into this while generating synthetic clinical data with TVAE, for an MSc dissertation that
I'm now turning into a paper. Rare categories never appeared in my generated data, and after looking
into it I think the cause might be in how `sample()` turns the decoder output back into a table. I
could be wrong, I'm still learning the codebase, but I wanted to share what I found in case it's
useful, and a clarification would help me a lot with the write-up.

TVAE **trains** the one-hot blocks as logits, with `cross_entropy` on the raw
decoder output, but `sample()` applies `tanh` to the whole output and hands it to
`inverse_transform`, which resolves each block with `argmax`. A class that is never the largest entry
for any row is therefore never generated, whatever probability the model assigned to it.
`CTGAN._apply_activate` in the same repository does draw from those blocks, with a Gumbel-softmax.

## What I observed

A discrete column with three classes at 90 / 9 / 1 per cent, 5,000 rows, `epochs=300`. The script is
in "Steps to reproduce" below, and this is its output:

| class | real | decoder believes | library output | after drawing |
|---|---|---|---|---|
| A | 90.26 | 92.76 | 97.06 | 92.60 |
| B | 8.72 | 6.25 | 2.94 | 6.46 |
| C | 1.02 | 0.98 | **0.00** | 0.94 |

`decoder believes` is the mean of `softmax(raw decoder output)` over the same 5,000 rows, so it is
what the model puts on each class before anything is decoded.

Reading the last row: the model puts 0.98 per cent on the rare class, against a true 1.02, and the
generated table contains **no row at all** for it. On three training seeds the rare class comes out
at 0.00, 0.02 and 0.00 per cent.

The middle class shows the same thing without disappearing: the model puts 6.25 per cent on it and
the generated table has 2.94, so it loses more than half of what the model believed. None of that
mass is lost, it is added to the dominant class, which goes from 92.76 believed to 97.06 written.

### The same draw decoded two ways

I reproduced the four steps of `sample()` by hand, keeping the raw decoder output, so that both
decodings start from exactly the same 5,000 noise vectors. The only difference between the two
columns on the right of the table is how each one-hot block is resolved.

Drawing from each block reproduces `decoder believes` on all three classes, to within a quarter of
a point. Taking the largest entry does not.

### Where the largest entry is taken

`ctgan/synthesizers/tvae.py`, in `sample()`, applies `tanh` to the whole vector and draws nothing:

```python
fake, sigmas = self.decoder(noise)
fake = torch.tanh(fake)
...
return self.transformer.inverse_transform(data, sigmas.detach().cpu().numpy())
```

`ctgan/data_transformer.py`, for a discrete column, hands the block to RDT's `OneHotEncoder`:

```python
def _inverse_transform_discrete(self, column_transform_info, column_data):
    ohe = column_transform_info.transform
    data = pd.DataFrame(column_data, columns=list(ohe.get_output_sdtypes()))
    return ohe.reverse_transform(data)[column_transform_info.column_name]
```

whose `_reverse_transform` resolves it by position of the maximum:

```python
indices = np.argmax(data, axis=1)
```

and for a continuous column the mode indicator is resolved the same way, here in `ctgan` itself:

```python
data[data.columns[1]] = np.argmax(column_data[:, 1:], axis=1)
```

### Training and sampling do not treat these blocks the same way

This is the part that makes me think it is an oversight rather than a choice. `_loss_function` walks
`output_info` and does two different things depending on the activation:

```python
def _loss_function(recon_x, x, sigmas, mu, logvar, output_info, factor):
    st = 0
    loss = []
    for column_info in output_info:
        for span_info in column_info:
            if span_info.activation_fn != 'softmax':
                ed = st + span_info.dim
                std = sigmas[st]
                eq = x[:, st] - torch.tanh(recon_x[:, st])
                loss.append((eq**2 / 2 / (std**2)).sum())
                loss.append(torch.log(std) * x.size()[0])
                st = ed

            else:
                ed = st + span_info.dim
                loss.append(
                    cross_entropy(
                        recon_x[:, st:ed], torch.argmax(x[:, st:ed], dim=-1), reduction='sum'
                    )
                )
                st = ed

    assert st == recon_x.size()[1]
    KLD = -0.5 * torch.sum(1 + logvar - mu**2 - logvar.exp())
    return sum(loss) * factor / x.size()[0], KLD / x.size()[0]
```

So during training a softmax span is treated as **logits**, with no `tanh` applied to it. At sampling
time the same span gets `tanh` and then `argmax`. The metadata that distinguishes the two kinds of
span, `span_info.activation_fn`, is read by the loss and ignored by `sample()`.

### The sibling model in the same repository does draw

`ctgan/synthesizers/ctgan.py`:

```python
def _apply_activate(self, data):
    """Apply proper activation function to the output of the generator."""
    data_t = []
    st = 0
    for column_info in self._transformer.output_info_list:
        for span_info in column_info:
            if span_info.activation_fn == 'tanh':
                ed = st + span_info.dim
                data_t.append(torch.tanh(data[:, st:ed]))
                st = ed
            elif span_info.activation_fn == 'softmax':
                ed = st + span_info.dim
                transformed = self._gumbel_softmax(data[:, st:ed], tau=0.2)
                data_t.append(transformed)
                st = ed
    return torch.cat(data_t, dim=1)
```

Both models carry the same `output_info_list`. CTGAN reads the activation and samples the softmax
spans; TVAE applies `tanh` to everything.

### detail about the `tanh`

`tanh` is monotone, so it does not change which entry is the largest and it is not the cause of what
I report above. But it does matter for any correction: a softmax taken **after** `tanh` is computed
on scores compressed into [-1, 1] and is close to uniform whatever the model believed. A fix would
have to read the block's distribution from the raw decoder output, before the squashing.

## Why this might not show up in tests

I had a look at the tests and could not find one that would catch this, though I may have missed
something. The integration test module says as much in [its own docstring](https://github.com/sdv-dev/CTGAN/blob/main/tests/integration/synthesizer/test_tvae.py#L4-L10)

Going through the four integration tests in `tests/integration/synthesizer/test_tvae.py`:
`test_drop_last_false` fits on `{'1': ['a', 'b', 'c'] * 150, '2': ['a', 'b', 'c'] * 150}` and asserts that at least 95 of 100
sampled rows have `row['1'] == row['2']`. Both columns are perfectly balanced, and the assertion
counts agreement between the two columns rather than proportions, so a model that only ever produced
`('a', 'a')` would score 100 on it. `test__loss_function` uses two numeric columns.
`test_fixed_random_seed` and `test_tvae_save` use `np.random.choice(['a', 'b', 'c'], 100)`, balanced
again, and assert reproducibility, shape, column names, dtypes and the loss log.

So no test fits an imbalanced categorical column, and none asserts that the categories present in the
output, or their proportions, match the training data, which is where this would show. The two unit
tests in `tests/unit/synthesizer/test_tvae.py` cover the constructor's defaults and the verbose
progress bar.

It would also be easy to miss on a balanced column: there the winning entry changes from row to
row, so the proportions come out roughly right and nothing looks wrong. What makes it visible is a
class rare enough to be the largest entry for no row at all.

## Related issues

sdv-dev/SDV#248: "For categorical column highly imbalanced categories are being lost in the generated data", on categories at about 1 per cent. It was answered and closed on the choice of RDT transformer, which acts on how a category is encoded before training, so as far as I can tell it is a different step from the one I'm describing here. I may be missing a connection between the two.

## Steps to reproduce

```python
import numpy as np
import pandas as pd
import torch
from ctgan import TVAE

RNG = np.random.default_rng(0)
N = 5000
cat = RNG.choice(['A', 'B', 'C'], size=N, p=[0.90, 0.09, 0.01])
which = RNG.choice([0, 1, 2], size=N, p=[0.60, 0.30, 0.10])
num = RNG.normal(np.array([0.0, 2.5, 5.0])[which], 1.0)
real = pd.DataFrame({'cat': cat, 'num': num})

torch.manual_seed(0)
np.random.seed(0)
tvae = TVAE(epochs=300)
tvae.fit(real, discrete_columns=['cat'])

# the four steps of sample(), by hand, keeping the raw decoder output
torch.manual_seed(0)
noise = torch.randn(N, tvae.embedding_dim).to(tvae._device)
tvae.decoder.eval()
with torch.no_grad():
    raw, sigmas = tvae.decoder(noise)
raw, sigmas = raw.cpu().numpy(), sigmas.detach().cpu().numpy()


def softmax(x):
    e = np.exp(x - x.max(axis=1, keepdims=True))
    return e / e.sum(axis=1, keepdims=True)


block = slice(0, 3)                      # the one-hot of 'cat'
p = softmax(raw[:, block])

# what the library does
shipped = tvae.transformer.inverse_transform(np.tanh(raw), sigmas)

# the same vector, with that one block drawn from its own distribution
rng = np.random.default_rng(0)
drawn_vector = np.tanh(raw)
drawn_vector[:, block] = np.eye(3)[np.array([rng.choice(3, p=row) for row in p])]
drawn = tvae.transformer.inverse_transform(drawn_vector, sigmas)

classes = list(tvae.transformer._column_transform_info_list[0].transform.dummies)


def rate(frame):
    return frame['cat'].value_counts(normalize=True).reindex(classes).fillna(0) * 100


print(pd.DataFrame({'real': rate(real),
                    'decoder believes': p.mean(axis=0) * 100,
                    'library output': rate(shipped),
                    'after drawing': rate(drawn)}, index=classes).round(2))
```

## What I'd expect

If I'm reading this correctly, `sample()` could apply the activation each span declares, the way the
generator of CTGAN already does, rather than `tanh` to everything. Something like:

```python
# in TVAE.sample(), replacing  fake = torch.tanh(fake)
fake = self._apply_activate(fake)
```

with `_apply_activate` mirroring the one in `ctgan.py`, except that for a softmax span a plain
categorical draw is enough at sampling time, since nothing is being differentiated there:

```python
def _apply_activate(self, data):
    data_t = []
    st = 0
    for column_info in self.transformer.output_info_list:
        for span_info in column_info:
            ed = st + span_info.dim
            if span_info.activation_fn == 'tanh':
                data_t.append(torch.tanh(data[:, st:ed]))
            else:
                probs = torch.softmax(data[:, st:ed], dim=1)
                picked = torch.multinomial(probs, num_samples=1)
                data_t.append(torch.zeros_like(probs).scatter_(1, picked, 1.0))
            st = ed
    return torch.cat(data_t, dim=1)
```

I've been running the equivalent of this outside the library and it behaves as expected. That said,
I'm not sure it is the right fix, and it does change the output of a released model, so it may well
be a decision for you rather than a patch to take. Two questions I'd genuinely like the answer to:

1. Is the deterministic decoding intended for TVAE, given that CTGAN samples the same blocks?
2. If it isn't, would you want this behind an argument rather than as the default?

I'm happy to open a PR with tests if that would help, and I have the full output and the notebook if
any of it is useful. Thanks a lot for the library, and for reading this far.
